tests/test_profiler.py

```diff
@@ -316,3 +316,21 @@ def test_pytorch_profiler_nested_emit_nvtx(tmpdir):
         gpus=1,
     )
     trainer.fit(model)
+
+
+@pytest.mark.parametrize("cls", (SimpleProfiler, AdvancedProfiler, PyTorchProfiler))
+def test_profiler_teardown(tmpdir, cls):
+    """
+    This test checks if profiler teardown method is called when trainer is exiting.
+    """
+    profiler = cls(output_filename=os.path.join(tmpdir, "profiler.txt"))
+
+    model = BoringModel()
+    trainer = Trainer(
+        default_root_dir=tmpdir,
+        fast_dev_run=True,
+        profiler=profiler,
+    )
+    trainer.fit(model)
+
+    assert profiler.output_file.closed
```

Comment:
one thing i'm concerned about is if someone does this:
```
trainer = Trainer(..., profiler=Profiler(...))
trainer.fit(module)
trainer.test(module)
```
with the setup and teardown hooks, will the `test` call overwrite the results from the `fit` call?

do we want to approach this with the proposal from #6420 and introduce additional hooks into the profiler API?

if you think that would bloat the profiler API, we could add some handling to include the stage information in the filename to avoid collisions, and then add versioning on top of that so we don't silently overwrite things

Reply:
Good point!

> do we want to approach this with the proposal from #6420 and introduce additional hooks into the profiler API?

I think I'd rather let other people chime in on that proposal instead of implementing it already for this. We also need to move fast so that would be more overhead for reviewers

So for now, we can take the `stage` argument in both `setup` and `teardown`, and do as you propose:

> include the stage information in the filename to avoid collisions, and then add versioning on top of that so we don't silently overwrite things

Reply:
sgtm - i agree #6420 shouldn't block this and if that's agreed on, we'll need to audit and fix systematically anyways.